icicle_v3/backend/cpu/src/curve/cpu_msm.hpp

```diff
@@ -0,0 +1,667 @@
+#pragma once
+
+#include <thread>
+#include <atomic>
+#include <string>
+#include <iostream>
+#include <fstream>
+
+#include "icicle/errors.h"
+#include "icicle/runtime.h"
+#include "icicle/config_extension.h"
+#include "icicle/curves/projective.h"
+#include "icicle/curves/curve_config.h"
+#include "icicle/msm.h"
+#include "tasks_manager.h"
+
+using namespace icicle;
+
+#ifdef DUMMY_TYPES // for testing
+using A = DummyP;
+using P = DummyP;
+using scalar_t = DummyScalar;
+#else
+using A = curve_config::affine_t;
+using P = curve_config::projective_t;
+using scalar_t = curve_config::scalar_t;
+#endif
+
+/**
+ * @class EcAddTask
+ * @brief class deriving `TaskBase` to allow use of tasks_manager.h for MSM
+ * (The task to be executed is an elliptic curve addition).
+ */
+template <typename A, typename P>
+class EcAddTask : public TaskBase
+{
+public:
+  /**
+   * @brief constructor for the task ensuring Ps are zeros and other fields are init to invalid values (to avoid
+   * falsely handling a result at the start). The execution adds Ps m_p1,m_p2 and stores the result in m_p1.
+   */
+  EcAddTask() : TaskBase(), m_p1(P::zero()), m_p2(P::zero()), m_return_idx(-1) {}
+
+  /**
+   * @brief Function to be executed by the tasks manager. It can be configured according to the value odf
+   */
+  void execute()
+  {
+    switch (m_p2_config) {
+    case ADD_P1_P2_BY_VALUE:
+      m_p1 = m_p1 + m_p2;
+      return;
+    case ADD_P1_AND_P2_PER:
+      m_p1 = m_p1 + *m_p2_Per;
+      return;
+    case ADD_P1_AND_AFFINE_P1:
+      m_p1 = m_p1 + m_p2_affine;
+      return;
+    }
+  }
+
+  /**
+   * @brief Set up phase 1 addition between a new base and the target bucket.
+   * @param bucket - bucket to be added. It isn't passed as Per as the bucket's value can be changed during
+   * addition execution.
+   * @param base - affine base to be added. Passed as a Per as the input is constant and it saves copying.
+   * @param negate_affine - flag to indicate that the base needs to be subbed instead of added.
+   * @param is_montgomery - flag to indicate that the base is in Montgomery form and first needs to be converted.
+   */
+  void set_phase1_addition_with_affine(const P& bucket, const A base, int bucket_idx)
```

Comment:
why base passed by value?

Reply:
Passed by value due to intermediate calculations done in bucket_accumulator which are stored in a local variable - meaning there is no valid address to be passed 
there are todos to make the intermediate calcuations as preprocessing which would also allow to pass the affine point by reference, but it will be added in a future update